### PR TITLE
Fix use of CMake build types

### LINKED ISSUE #6419: RelWithDebInfo build type
Build type `Test` in CMakeLists.txt corresponds to default CMake build type `RelWithDebInfo` (release with debug on).

There is also a section that sets up `-O0` for `Debug`, `-O3` for release and so on, which should not be necessary.

`Test` build type is tied to `--test` flag in the script, but `RelWithDebInfo` can work just as fine.

### PR DESCRIPTION
**Pull request by ppenzin**
Debug, Release, and DebWithRelInfo set optimization levels and also
toggle debug generation.

Fixes #6419

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -402,19 +402,12 @@ if(CMAKE_BUILD_TYPE STREQUAL Debug)
         -D_DEBUG=1 # for PAL
         -DDBG_DUMP=1
     )
-elseif(CMAKE_BUILD_TYPE STREQUAL Test)
+elseif(CMAKE_BUILD_TYPE STREQUAL RelWithDebInfo)
     add_definitions(
         -DENABLE_DEBUG_CONFIG_OPTIONS=1
     )
-    add_compile_options(-g)
```
Comment (rhuanjl) [inline CMakeLists.txt:409]: Was removing this -g option intentional?

Comment (ppenzin) [reply CMakeLists.txt:409]: Yep, `RelWithDebInfo` adds it by default.

**Comment by ppenzin** [comment]
I might need to bump opt level for the build type (and remove -DNDEBUG), let's see if that is why CI failed.

**Comment by ppenzin** [comment]
I ended up bumping opt level and removing -DNDEBUG, to match what `Test` mode was doing.

**Comment by rhuanjl** [comment]
Hmm - macOS system ICU has missing functions that CC expects. And if I try using the version of ICU downloaded by build.sh (normally embeded) but building with cmake it doesn't get embedded and it fails to find it at run time.

I can use the cmake build to build a no-icu version, if I try testing it numerous tests fail - presumably due to no-icu, though some of the failures are segfaults which seems surprising to me if it's just no-icu.

**Pull request merged**